pkg/prom/cleaner.go

```diff
@@ -0,0 +1,258 @@
+		done:            make(chan bool),
+	}
+
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	// If the caller hasn't set a ticker, create one with the default period
+	if c.ticker == nil {
+		c.ticker = time.NewTicker(DefaultCleanupPeriod)
+	}
+
+	go c.run()
+	return c
+}
+
+// Get storage directories used for each ManagedInstance
+func (c *WALCleaner) getManagedStorage(instances map[string]instance.ManagedInstance) map[string]bool {
+	out := make(map[string]bool)
+
+	for _, inst := range instances {
+		out[inst.StorageDirectory()] = true
+	}
+
+	return out
+}
+
+// Get all Storage directories under walDirectory
+func (c *WALCleaner) getAllStorage() []string {
+	var out []string
+
+	_ = filepath.Walk(c.walDirectory, func(p string, info os.FileInfo, err error) error {
+		if err != nil {
+			if os.IsNotExist(err) {
+				// The root WAL directory doesn't exist. Maybe this Agent isn't responsible for any
+				// instances yet. Log at debug since this isn't a big deal. We'll just try to crawl
+				// the direction again on the next periodic run.
+				level.Debug(c.logger).Log("msg", "WAL storage path does not exist", "path", p, err)
+			} else {
+				// Just log any errors traversing the WAL directory. This will potentially result
+				// in a WAL (that has incorrect permissions or some similar problem) not being cleaned
+				// up. This is  better than preventing *all* other WALs from being cleaned up.
+				discoveryError.WithLabelValues(p).Inc()
+				level.Warn(c.logger).Log("msg", "unable to traverse WAL storage path", "path", p, "err", err)
+			}
+		} else if info.IsDir() && filepath.Dir(p) == c.walDirectory {
+			// Single level below the root are instance storage directories (including WALs)
+			out = append(out, p)
+		}
+		return nil
+	})
+
+	return out
+}
+
+// Get the full path of storage directories that aren't associated with an active instance
+// and haven't been written to within a configured duration (usually several hours or more).
+func (c *WALCleaner) getAbandonedStorage(all []string, managed map[string]bool, now time.Time) []string {
+	var out []string
+
+	for _, dir := range all {
+		if !managed[dir] {
+			walDir := wal.SubDirectory(dir)
+			mtime, err := c.walOperations.lastModified(walDir)
+			if err != nil {
+				segmentError.WithLabelValues(dir).Inc()
+				level.Warn(c.logger).Log("msg", "unable to find segment mtime of WAL", "name", dir, "err", err)
+				continue
+			}
+
+			diff := now.Sub(mtime)
+			if diff > c.minAge {
+				// The last segment for this WAL was modified more then $minAge (positive number of hours)
+				// in the past. This makes it a candidate for deletion since it's also not associated with
+				// any Instances this agent knows about.
+				out = append(out, dir)
+			}
+
+			level.Debug(c.logger).Log("msg", "abandoned WAL", "name", dir, "mtime", mtime, "diff", diff)
+		} else {
+			level.Debug(c.logger).Log("msg", "active WAL", "name", dir)
+		}
+	}
+
+	return out
+}
+
+func (c *WALCleaner) run() {
```

Comment:
Can we add a histogram to track the time it takes to perform a clean cycle? I'm interested in how long it will take for significant deployments of the Agent.

Reply:
Good idea